app/src/main/java/rick_and_morty/ui/characters/CharactersViewModel.kt

```diff
@@ -20,30 +19,35 @@ class CharactersViewModel @Inject constructor(
     private var initialLoad: Boolean = true
     private var page = 1;
 
-    private var _characters = MutableStateFlow<List<CharacterResultsDto>>(emptyList())
-    val characters: StateFlow<List<CharacterResultsDto>> = _characters
+    private var _characters = MutableStateFlow(UIState())
+    val characters: StateFlow<UIState> = _characters
+
+    private var characterList = mutableListOf<CharacterResultsDto>()
+
 
     init {
         getCharacters()
         initialLoad = false
     }
 
     fun getCharacters() {
+        _characters.update { it.copy(isLoading = true) }
         if (!initialLoad) {
             viewModelScope.launch {
-                try {
-                    val getCharacter = characterRepository.getCharacters(page)
-                    _characters.update {
-                        it + getCharacter
-                    }
-                        page += 1;
-                    }
-                    catch(error: Exception) {
-                        Log.e("GetCharacterError", error.toString())
+                val getCharacter = characterRepository.getCharacters(page)
+                _characters.update {
+                    when(getCharacter){
+                        is BusEvent.Success -> { getCharacter.list.let { characterList.addAll(it) }
+                            it.copy(isLoading = false, isSuccess = characterList)
+                        }
+                        is BusEvent.Error -> it.copy(isLoading = false, isFailure = true, failureMessage = getCharacter.exception)
                     }
```

Comment:
1. When this is identified as a `Success` event, please use the values stored in the event instead of `getCharacter`. It should be something like `it.list`.
2. You could also use `buildList { ... }` (https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-list.html) to make data mutable and immutable again.